### PR TITLE
SH-469 - Unit to Course functionality added

### PR DESCRIPTION
**Pull request by rheafernandes**

### CODE REVIEW

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 1]
```diff
@@ -23,10 +23,12 @@
 
 import static org.sunbird.common.models.util.JsonKey.EKSTEP_BASE_URL;
 import static org.sunbird.common.models.util.ProjectUtil.getConfigValue;
-import static org.sunbird.common.responsecode.ResponseCode.SERVER_ERROR;
+import static org.sunbird.common.responsecode.ResponseCode.CLIENT_ERROR;
+import static org.sunbird.common.models.util.JsonKey.CONTENT_PROPS_TO_ADD;
 
 public class CourseManagementActor extends BaseActor {
     private static ObjectMapper mapper = new ObjectMapper();
+    private static List<String> metadataToBeAdded = Arrays.stream(getConfigValue(CONTENT_PROPS_TO_ADD).split(",")).map(String::trim).collect(Collectors.toList());
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:31]: Add empty list incase config property is not available. Also check for empty list while using **metadataToBeAdded** for filtration. If this list is empty throw client exception.

Comment (amitpriyadarshi) [reply course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:31]: Also add one unit test case for nullity and empty check.

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 2]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+                            .body(mapper.writeValueAsString(generateUpdateHierarchyRequest(request, createResponse)))
+                            .asString();
+            if (null != updateResponse) {
+                Response response = mapper.readValue(updateResponse.getBody(), Response.class);
+                ProjectLogger.log(
+                        "Sized: CourseManagementActor:handleHierarchyData : size of response : "
+                                + updateResponse.getBody().getBytes().length,
+                        LoggerEnum.INFO);
+                if (!StringUtils.equalsIgnoreCase(response.getResponseCode().name(), ResponseCode.OK.name()))
+                    ProjectCommonException.throwClientErrorException(
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:152]: Add log statement and print error message.

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 3]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+            String url = getConfigValue(EKSTEP_BASE_URL) + "/content/v3/hierarchy/update";
+            HttpResponse<String> updateResponse =
+                    Unirest.patch(url)
+                            .headers(headers)
+                            .body(mapper.writeValueAsString(generateUpdateHierarchyRequest(request, createResponse)))
+                            .asString();
+            if (null != updateResponse) {
+                Response response = mapper.readValue(updateResponse.getBody(), Response.class);
+                ProjectLogger.log(
+                        "Sized: CourseManagementActor:handleHierarchyData : size of response : "
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:148]: remove this log

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 4]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
                 throw new ProjectCommonException(
                         ResponseCode.CLIENT_ERROR.getErrorCode(),
                         ResponseCode.CLIENT_ERROR.getErrorMessage(),
-                        SERVER_ERROR.getResponseCode());
+                        CLIENT_ERROR.getResponseCode());
             }
         }
     }
 
+    private void handleHierarchyData(Request request, Response createResponse, Map<String, String> headers) throws Exception {
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:137]: Just pass identifier of course

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 5]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+                        + mapper.writeValueAsString(updateRequest),
+                LoggerEnum.INFO.name());
+        return updateRequest;
+    }
+
+    private void getRecursiveHierarchyRequest(String parentId, List<Map<String, Object>> children, Map<String, Object> nodesModified,
+                                              Map<String, Object> hierarchy, Boolean root) {
+        children.forEach(child -> {
+            String identifier = (String) child.get(SunbirdKey.IDENTIFIER);
+            if(!child.containsKey(SunbirdKey.VISIBILITY))
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:187]: check child contains list of 4 properties - identifier, mimetype, visibility, contentType

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 6]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+                    && !StringUtils.equalsIgnoreCase((String) child.get(SunbirdKey.VISIBILITY), SunbirdKey.VISIBILITY_DEFAULT))
+                getRecursiveHierarchyRequest(identifier,
+                        (List<Map<String, Object>>) child.getOrDefault(SunbirdKey.CHILDREN, new ArrayList<Map<String, Object>>()),
+                        nodesModified, hierarchy, false);
+        });
+    }
+
+    private Map<String, Object> getNodeModifiedMap(Map<String, Object> metadata) {
+        metadata.put(SunbirdKey.CONTENT_TYPE, "CourseUnit");
+        return new HashMap<String, Object>() {{
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:209]: add metaData.get("identifier") as origin into new cleanedup metadata.
"originData": "{\"name\":\"Ab103\"}

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 7]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+                                              Map<String, Object> hierarchy, Boolean root) {
+        children.forEach(child -> {
+            String identifier = (String) child.get(SunbirdKey.IDENTIFIER);
+            if(!child.containsKey(SunbirdKey.VISIBILITY))
+                throw new ProjectCommonException(
+                        ResponseCode.CLIENT_ERROR.getErrorCode(),
+                        "visibility is a mandatory parameter for content with id: " + identifier,
+                        CLIENT_ERROR.getResponseCode());
+            if (StringUtils.equalsIgnoreCase((String) child.get(SunbirdKey.VISIBILITY), SunbirdKey.VISIBILITY_PARENT)) {
+                identifier = System.currentTimeMillis() + "";
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:193]: rename **identifier** as **code**

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 8]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+                nodesModified.put(identifier, getNodeModifiedMap(child));
+            }
+            if (MapUtils.isEmpty(((Map<String, Object>) hierarchy.getOrDefault(parentId, new HashMap<String, Object>()))))
+                hierarchy.put(parentId, getNodeHierarchyMap(root));
+            ((List<String>) ((Map<String, Object>) hierarchy.get(parentId)).get(SunbirdKey.CHILDREN)).add(identifier);
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:198]: ** identifier** to **code**

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 9]
```diff
@@ -126,9 +129,101 @@ private void createCourse(Request request) throws Exception {
...
+            if(!child.containsKey(SunbirdKey.VISIBILITY))
+                throw new ProjectCommonException(
+                        ResponseCode.CLIENT_ERROR.getErrorCode(),
+                        "visibility is a mandatory parameter for content with id: " + identifier,
+                        CLIENT_ERROR.getResponseCode());
+            if (StringUtils.equalsIgnoreCase((String) child.get(SunbirdKey.VISIBILITY), SunbirdKey.VISIBILITY_PARENT)) {
+                identifier = System.currentTimeMillis() + "";
+                nodesModified.put(identifier, getNodeModifiedMap(child));
+            }
+            if (MapUtils.isEmpty(((Map<String, Object>) hierarchy.getOrDefault(parentId, new HashMap<String, Object>()))))
```
Comment (amitpriyadarshi) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:196]: Here default value is not required.

File: course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java [hunk 10]
```diff
@@ -122,11 +122,42 @@ private void createCourse(Request request) throws Exception {
...
         }
     }
 
+    private void handleHierarchyData(Request request, String identifier, Map<String, String> headers) throws Exception {
+        if (request.getRequest().containsKey(SunbirdKey.HIERARCHY)) {
+            String url = getConfigValue(EKSTEP_BASE_URL) + "/content/v3/hierarchy/update";
+            HttpResponse<String> updateResponse =
+                    Unirest.patch(url)
+                            .headers(headers)
+                            .body(mapper.writeValueAsString(new HierarchyGenerationHelper().generateUpdateHierarchyRequest(request, identifier)))
```
Comment (maheshkumargangula) [inline course-mw/course-actors-common/src/main/java/org/sunbird/learner/actors/course/CourseManagementActor.java:138]: HierarchyGenerationHelper - Create a Global variable and use it. Don't create a new variable for each request.
